jdtls.ext/com.microsoft.jdtls.ext.core/src/org/eclipse/jdt/internal/jarpackager/JarPackageUtil.java

```diff
@@ -0,0 +1,378 @@
+		}
+	}
+
+	/**
+	 * Write the entry to the destinationPath of the given JarOutputStream.
+	 * Extracted from org.eclipse.jdt.internal.ui.jarpackagerfat.JarWriter4
+	 *
+	 * @param	zipEntry					the jar entry to write
+	 * @param	zipFile						the zipfile to extract
+	 * @param	path						the destinationPath in the jar file
+	 * @param	areDirectoryEntriesIncluded	Tells whether directory entries are added to the jar
+	 * @param	isCompressed				whether the jar is compressed or not
+	 * @param	fJarOutputStream			the destination JarOutputStream
+	 * @param	fDirectories				the temporary set saves existing directories
+	 *
+	 * @throws	IOException					If an I/O error occurred
+	 *
+	 * @since	1.14
+	 *
+	 */
+	private static void addZipEntry(ZipEntry zipEntry, ZipFile zipFile, String path,
+			boolean areDirectoryEntriesIncluded, boolean isCompressed,
+			JarOutputStream fJarOutputStream, Set<String> fDirectories) throws IOException {
+		if (areDirectoryEntriesIncluded)
+			addDirectories(path, fJarOutputStream, fDirectories);
+
+		JarEntry newEntry= new JarEntry(path.replace(File.separatorChar, '/'));
+
+		if (isCompressed)
+			newEntry.setMethod(ZipEntry.DEFLATED);
+			// Entry is filled automatically.
+		else {
+			newEntry.setMethod(ZipEntry.STORED);
+			newEntry.setSize(zipEntry.getSize());
+			newEntry.setCrc(zipEntry.getCrc());
+		}
+
+		long lastModified= System.currentTimeMillis();
+
+		// Set modification time
+		newEntry.setTime(lastModified);
+
+		addEntry(newEntry, zipFile.getInputStream(zipEntry), fJarOutputStream);
+	}
+
+	/**
+	 * Creates the directory entries for the given path and writes it to the current archive.
+	 * Extracted from org.eclipse.jdt.ui.jarpackager.JarWriter3
+	 *
+	 * @param	destPath				the path to add
+	 * @param	fJarOutputStream		the destination JarOutputStream
+	 * @param	fDirectories			the temporary set saves existing directories
+	 *
+	 * @throws	IOException				if an I/O error has occurred
+	 *
+	 * @since	1.14
+	 */
+	private static void addDirectories(String destPath, JarOutputStream fJarOutputStream, Set<String> fDirectories) throws IOException {
+		String path= destPath.replace(File.separatorChar, '/');
+		int lastSlash= path.lastIndexOf('/');
+		List<JarEntry> directories= new ArrayList<>(2);
+		while (lastSlash != -1) {
+			path= path.substring(0, lastSlash + 1);
+			if (!fDirectories.add(path))
+				break;
+
+			JarEntry newEntry= new JarEntry(path);
+			newEntry.setMethod(ZipEntry.STORED);
+			newEntry.setSize(0);
+			newEntry.setCrc(0);
+			newEntry.setTime(System.currentTimeMillis());
+			directories.add(newEntry);
+
+			lastSlash= path.lastIndexOf('/', lastSlash - 1);
+		}
+
+		for (int i= directories.size() - 1; i >= 0; --i) {
+			fJarOutputStream.putNextEntry(directories.get(i));
+		}
+	}
+
+	/**
+	 * Write the single file to the JarOutputStream.
+	 * Extracted from org.eclipse.jdt.internal.ui.jarpackagerfat.JarWriter4
+	 *
+	 * @param	file						the file to write
```

Comment:
Again, please pay attention to the format of the JavaDoc. Take this as an example https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html